### PR TITLE
[Doc Day]add new tra example agent

### PR DESCRIPTION
**Pull request by dexterju**
**Patch description**
This adds a new example agent for how to use torch ranker agent.

**Testing steps**
```
python  examples/train_model.py -m examples/tra  -t convai2 -mf /tmp/test -bs 5
```
And 
```
 python examples/train_model.py -m examples/tra  -t convai2 -mf /tmp/test -bs 5 --validation-every-n-secs 5
```

### CODE REVIEW

File: parlai/agents/examples/tra.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import torch
+from torch import nn
+
+
+class ExampleBagOfWordsModel(nn.Module):
+    def __init__(self, opt, dictionary):
+        super().__init__()
+        self.hidden_dim = opt.get('hidden_dim', 512)
+        self.dict = dictionary
+        self.context_encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: you could even use the same embedding bag for even simpler.

File: parlai/agents/examples/tra.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        self.candidate_encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
+
+    def forward_candidates(self, cand_vecs):
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: docstring

File: parlai/agents/examples/tra.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+"""
+This is an example how to extend torch ranker agent and use it for your own purpose.
+In this example, we will just use a simple bag of words model.
+"""
+from parlai.core.torch_ranker_agent import TorchRankerAgent
+import torch
+from torch import nn
+
+
+class ExampleBagOfWordsModel(nn.Module):
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: docstring describing the general thing

File: parlai/agents/examples/tra.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        super().__init__()
+        self.hidden_dim = opt.get('hidden_dim', 512)
+        self.dict = dictionary
+        self.context_encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
+        self.candidate_encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
+
+    def forward_candidates(self, cand_vecs):
+        return self.candidate_encoder.forward(cand_vecs)
+
+    def forward_context(self, batch):
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: docstring

File: parlai/agents/examples/tra.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        context_encodings = self.forward_context(batch)
+        if context_encodings.dim() != cand_encs.dim():
+            # it means in eval mode
+            return torch.sum(
+                context_encodings.unsqueeze(1).expand_as(cand_encs) * cand_encs, 2
+            )
+        return context_encodings.mm(cand_encs.t())
+
+
+class TraAgent(TorchRankerAgent):
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: docstring

File: parlai/agents/examples/tra.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add CLI args.
+        """
+        TorchRankerAgent.add_cmdline_args(argparser)
+        arg_group = argparser.add_argument_group('ExampleBagOfWordsModel Arguments')
+        arg_group.add_argument('--hidden-dim', type=int, default=512)
+
+    def __init__(self, opt, shared=None):
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: you can drop this entire thing

File: parlai/agents/examples/tra.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        super().__init__(opt, shared)
+        self.opt = opt
+        return
+
+    def score_candidates(self, batch, cand_vecs, cand_encs=None):
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: Properly format the docstring

```
"""
A short, single-sentence, one-line description.

A longer descrpition containing any other information you need.
"""
```

File: parlai/agents/examples/tra.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        the likelihood that the candidate vector at that index is the
+        proper response. If `cand_encs` is not None (when we cache the
+        encoding of the candidate vectors), you may use these instead of
+        calling self.model on `cand_vecs`.
+        """
+        scores = self.model.forward(batch, cand_vecs, cand_encs)
+        return scores
+
+    def build_model(self):
+        """This function is required to build the model and assign to the
```
Comment (stephenroller) [inline parlai/agents/examples/tra.py]: Ditto on formatting.

File: parlai/agents/examples/tra.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        scores = self.model.forward(batch, cand_vecs, cand_encs)
+        return scores
+
+    def build_model(self):
+        """
+        This function is required to build the model and assign to the
+        object `self.model`.
+        """
+        self.model = ExampleBagOfWordsModel(self.opt, self.dict)
```
Comment (emilydinan) [inline parlai/agents/examples/tra.py]: nit: you don't need to assign `self.model` here. just `return ExampleBagOfWordsModel(self.opt, self.dict)`

File: parlai/agents/examples/tra.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        btz = cand_vecs.size(0)
+        if cand_encs is None:
+            if cand_vecs.dim() == 3:
+                # if dim = 3, btz *  num_candidates * seq_length
+                # In this case, we are using inline candidates
+                cand_vecs = cand_vecs.reshape(-1, cand_vecs.size(2))
+            cand_encs = self.forward_candidates(cand_vecs)
+            if cand_encs.size(0) != btz:
+                # Some cases, we could also use batch candidates,
+                # they will be size of btz * seq_length
```
Comment (emilydinan) [inline parlai/agents/examples/tra.py]: nit: elsewhere in the TRA code, we use the abbreviation "bsz" for "batchsize"

File: parlai/agents/examples/tra.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(self, opt, dictionary):
+        super().__init__()
+        self.hidden_dim = opt.get('hidden_dim', 512)
+        self.dict = dictionary
+        self.encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
+
+    def forward_candidates(self, cand_vecs):
+        """
+        This function will encode all the candidates for a batch. It can be different cases:
+        1. inline candidate: we have inline candidates provided for each sample, it usually happens during eval / test time
```
Comment (emilydinan) [inline parlai/agents/examples/tra.py]: this list isn't comprehensive. they are actually all described in the `_build_candidates` function docstring from TRA

File: parlai/agents/examples/tra.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            return torch.sum(
+                context_encodings.unsqueeze(1).expand_as(cand_encs) * cand_encs, 2
+            )
+        return context_encodings.mm(cand_encs.t())
+
+
+class TraAgent(TorchRankerAgent):
+    """
+    This subclass the torch ranker agent.
+    What you need to do:
```
Comment (emilydinan) [inline parlai/agents/examples/tra.py]: this docstring is a bit confusing as all of these things are done in this example agent. can you explain that this is a simple bag of words model?

File: parlai/agents/examples/tra.py (hunk 11 above)
Comment (dexterju) [reply parlai/agents/examples/tra.py]: ....What should I write here? Stephen asks me for a docstring here,

File: parlai/agents/examples/tra.py (hunk 12 above)
Comment (dexterju) [reply parlai/agents/examples/tra.py]: Nahn.... As the name of the class suggests? lol

Comment (stephenroller) [reply parlai/agents/examples/tra.py]:
```
Example subclass of TorchRankerAgent.

This particular implementation is a simple bag-of-words model, which
demonstrates the minimum implementation requirements to make a
new ranking model.
```

File: parlai/agents/examples/tra.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    def forward_candidates(self, cand_vecs):
+        """
+        This function will encode all the candidates for a batch. It can be different cases:
+        1. inline candidate: we have inline candidates provided for each sample, it usually happens during eval / test time
+        In this case, the cand_vecs is a 3D tensor, [btz, num_candidates, seq_length]
+        2. batch candidates, it usually happens during training, cand_vecs will be [btz, seq_length]
+        3. fixed, when you have a fixed set of candidates to use, cand_vecs will be [num_candidates, seq_length]
+        """
+        return self.encoder.forward(cand_vecs)
```
Comment (emilydinan) [inline parlai/agents/examples/tra.py]: also why do you explicitly need to call forward here?  if this function is the same for the context/candidates, then why do we need separate functions? just for clarity?

File: parlai/agents/examples/tra.py (hunk 11 above)
Comment (emilydinan) [reply parlai/agents/examples/tra.py]: """
Encodes the candidate responses.

Recall that candidates can be inline, batch, fixed, or created from the vocabulary. These candidates are built using the `_build_candidates` function in Torch Ranker Agent (add link here for description).
"""

File: parlai/agents/examples/tra.py (hunk 13 above)
Comment (dexterju) [reply parlai/agents/examples/tra.py]: Yes, I just did it for clarity purposes. I will remove it then!

**Comment by dexterju** [comment]
I think this is about all we need?

File: tests/test_examples.py [hunk 14]
```diff
@@ -46,6 +46,20 @@ def test_generation(self):
             "test token_acc = {}\nLOG:\n{}".format(test['token_acc'], stdout),
         )
 
+    @testing_utils.retry(ntries=3)
+    def test_repeater(self):
+        """
+        Test a simple TRA based bag-of-words model.
+        """
+        stdout, valid, test = testing_utils.train_model(
+            dict(task='integration_tests', model='examples/tra', eps=1, bs=100)
```
Comment (stephenroller) [inline tests/test_examples.py]: this won't work. you need to give full options, not short ones.

File: tests/test_examples.py [hunk 15]
```diff
@@ -46,6 +46,20 @@ def test_generation(self):
...
+        )
+        self.assertEqual(
+            test['exs'],
+            100,
+            'test examples = {}\nLOG:\n{}'.format(valid['exs'], stdout),
+        )
```
Comment (stephenroller) [inline tests/test_examples.py]: do you think there should be more checks than just "it did 100 examples"?

Comment (dexterju) [reply tests/test_examples.py]: I don't know what else should be checked? Like a certain performance?

Comment (dexterju) [reply tests/test_examples.py]: I added the test to check accuracy for test and validation as well.

**Comment by stephenroller** [comment]
okay let's land
